src/server/list_family.cc

```diff
@@ -853,6 +891,77 @@ OpResult<string> ListFamily::OpIndex(const OpArgs& op_args, std::string_view key
   return str;
 }
 
+OpResult<vector<uint32_t>> ListFamily::OpIndex(const OpArgs& op_args, std::string_view key,
+                                               std::string_view element, ArgSlice args_ext) {
+  OpResult<PrimeIterator> it_res = op_args.shard->db_slice().Find(op_args.db_cntx, key, OBJ_LIST);
+  if (!it_res.ok())
+    return it_res.status();
+
+  int direction = AL_START_HEAD;
+  int index = 0;
+  int count = -1;
+  int rank = 1;
+  int max_len = 0;
+  vector<uint32_t> matches;
+
+  for (size_t i = 0; i < args_ext.size(); i++) {
+    const auto& arg_v = args_ext[i];
+
+    if (arg_v == "RANK") {
+      if (!absl::SimpleAtoi(args_ext[i + 1], &rank)) {
+        return OpStatus::WRONG_TYPE;
+      } else if (rank < 0) {
+        rank = -rank;
+        direction = AL_START_TAIL;
+      } else if (rank == 0) {
+        return OpStatus::INVALID_NUMERIC_RESULT;
+      }
+    }
+    if (arg_v == "COUNT") {
+      if (!absl::SimpleAtoi(args_ext[i + 1], &count)) {
+        return OpStatus::WRONG_TYPE;
+      } else if (count < 0) {
+        return OpStatus::INVALID_NUMERIC_RESULT;
+      }
+    }
+    if (arg_v == "MAXLEN") {
+      if (!absl::SimpleAtoi(args_ext[i + 1], &max_len)) {
+        return OpStatus::WRONG_TYPE;
+      } else if (max_len < 0) {
+        return OpStatus::INVALID_NUMERIC_RESULT;
+      }
+    }
+  }
+
+  quicklist* ql = GetQL(it_res.value()->second);
+  quicklistIter* ql_iter = quicklistGetIterator(ql, direction);
+  quicklistEntry entry;
+
+  int matched = 0;
+  string str;
+
+  while (quicklistNext(ql_iter, &entry) && (max_len == 0 || index < max_len)) {
+    if (entry.value) {
+      str.assign(reinterpret_cast<char*>(entry.value), entry.sz);
+    } else {
+      str = absl::StrCat(entry.longval);
+    }
```

Comment:
Just as a suggestion. It'd be really cool if you could extend `container_utils::IterateList` to support reverse iteration (by adding an optional argument at the end).
This way you could use only it to implement the full logic without repeating low-level code.

Reply:
I'm `not` against it, maybe I can add this in the near future? Currently extending it without causing breakage would take some time.

Reply:
sure.